components/brave_wallet/browser/internal/orchard_block_scanner.cc

```diff
@@ -5,74 +5,75 @@
 
 #include "brave/components/brave_wallet/browser/internal/orchard_block_scanner.h"
 
+#include "base/threading/thread_restrictions.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/orchard_block_decoder.h"
+#include "brave/components/brave_wallet/browser/zcash/rust/orchard_decoded_blocks_bundle.h"
+
 namespace brave_wallet {
 
 OrchardBlockScanner::Result::Result() = default;
 
-OrchardBlockScanner::Result::Result(std::vector<OrchardNote> discovered_notes,
-                                    std::vector<OrchardNoteSpend> spent_notes)
+OrchardBlockScanner::Result::Result(
+    std::vector<OrchardNote> discovered_notes,
+    std::vector<OrchardNoteSpend> spent_notes,
+    std::unique_ptr<orchard::OrchardDecodedBlocksBundle> scanned_blocks)
     : discovered_notes(std::move(discovered_notes)),
-      spent_notes(std::move(spent_notes)) {}
-
-OrchardBlockScanner::Result::Result(const Result&) = default;
+      found_spends(std::move(spent_notes)),
+      scanned_blocks(std::move(scanned_blocks)) {}
 
+OrchardBlockScanner::Result::Result(OrchardBlockScanner::Result&&) = default;
 OrchardBlockScanner::Result& OrchardBlockScanner::Result::operator=(
-    const Result&) = default;
+    OrchardBlockScanner::Result&&) = default;
 
 OrchardBlockScanner::Result::~Result() = default;
 
-OrchardBlockScanner::OrchardBlockScanner(
-    const OrchardFullViewKey& full_view_key)
-    : decoder_(orchard::OrchardBlockDecoder::FromFullViewKey(full_view_key)) {}
+OrchardBlockScanner::OrchardBlockScanner(const OrchardFullViewKey& fvk)
+    : fvk_(fvk) {}
 
 OrchardBlockScanner::~OrchardBlockScanner() = default;
 
 base::expected<OrchardBlockScanner::Result, OrchardBlockScanner::ErrorCode>
 OrchardBlockScanner::ScanBlocks(
-    std::vector<OrchardNote> known_notes,
-    std::vector<zcash::mojom::CompactBlockPtr> blocks) {
+    const OrchardTreeState& tree_state,
```

Comment:
I have looked about and it looks like `tree_state` should be taken by value and moving it with `std::move`. This will avoid the deep copy in `DecodeBlocks` where we are doing something like this:

```cxx
  base::ranges::copy(tree_state.frontier,
                     std::back_inserter(prior_tree_state.frontier));
```

and we could be just moving it:
```cxx
prior_tree_state.frontier = std::move(tree_state.frontier);
```

What do you think?

Reply:
This is actually copies to the different container type, so we can't just move here
```
  base::ranges::copy(tree_state.frontier,
                     std::back_inserter(prior_tree_state.frontier));
```

Reply:
What is the type of `prior_tree_state.frontier`? Couldn't we just do:

```cxx
prior_tree_state.frontier = std::move(tree_state.frontier);
```

Reply:
prior_tree_state.frontier is ::rust::Vec<::std::uint8_t> while tree_state.frontier is std::vector<uint8_t>

Reply:
Ack